### DISCUSSION TITLE
Educational Applications of AI in Web Browsers

### DISCUSSION THREAD

**Discussion by AdamSobieski** [microsoft/semantic-kernel#108] [General]
Hello. I am interested in educational applications of AI in Web browsers.

I am looking forward to students being able to ask questions about and to engage in dialogues about Web documents, e.g., Wikipedia articles, and to do so with respect to PDF documents displayed in their Web browsers, e.g., arXiv, Google Scholar, Semantic Scholar, etc. Interestingly, these technologies could also benefit students by interoperating with their digital textbooks.

With respect to interoperations between AI and Web browsers, one can envision the page and scroll positions of documents as well as any content selections as being relevant to end-users’ questions and dialogues. One can also envision AI components being able to turn pages, scroll documents, and highlight selections of content while responding to end-users. Soon, AI systems may be able to open documents for end-users and subsequently turn pages, scroll to, and highlight relevant content while communicating.

I would like to request more information and documentation about how the document-related skills work. I am eager to learn more about these topics to be able to better brainstorm about educational applications of AI in Web browsers and about how new Web standards could be of use.

For starters: Can developers provide one or more documents' URLs and, potentially, content selectors to LLMs? Can developers ask questions while indicating for LLMs to consider one or more documents, or chapters or sections thereof, as being prioritized above others while answering those questions? Would this, instead, require uploading portions of documents before or during end-users' questions or dialogue moves? Thank you for any more information about LLMs and document-related skills!

**Comment by timlaverty** [comment]
Hi Adam, Great questions.  Some answers to get you started:

> Can developers provide one or more documents' URLs and, potentially, content selectors to LLMs? 

Yes, here's a prompt asked of -003:

```
summarize the article at this URL: https://en.wikipedia.org/wiki/RimWorld

A possible summary is:

RimWorld is a sci-fi simulation game that allows players to create and manage a colony of survivors on a distant planet. The game is influenced by the works of Isaac Asimov, Robert A. Heinlein, and Ursula K. Le Guin, and features a dynamic story generator that adapts to the players' actions and choices. The game also incorporates elements of psychology, ecology, combat, and social relations, as well as various scenarios and challenges.
```

> Can developers ask questions while indicating for LLMs to consider one or more documents, or chapters or sections thereof, as being prioritized above others while answering those questions? 

I think so, i.e., you can ask the AI to prioritize different content in its responses.  I believe sections and other docs would be fine.  This would be worth experimenting for speicfics though:

```
Write a 1 paragraph summary of this video game: https://en.wikipedia.org/wiki/RimWorld but prefer the content in https://rimworldwiki.com/wiki/Ideology_(DLC) because it's a later update to the core game.

Possible summary:

RimWorld is a sci-fi colony simulation game where players manage the lives, needs, and beliefs of a diverse group of survivors on a hostile planet. The game features a dynamic story generator that creates events and challenges based on the colonists' personalities, skills, and relationships. In the Ideology DLC, players can customize their own belief system or choose from various pre-made ones, such as transhumanism, or tribalism, and shape their colony's
```


> Would this, instead, require uploading portions of documents before or during end-users' questions or dialogue moves? 

You could include the docs & articles in your prompts.  Another thought is using a mix of native and semantic code to chunk your docs up, summarize pieces of them, and ask the AI to adjust priority based on those summaries.  Skills I think you might be interested in: HTTPSkill, Summarizer skill.  I suspect you will want to customize what is there, using your own custom prompts as you build our your app above.  I suspect though the inbox skills will get you through a POC and give you a feel for what you'd want to change / add.

HTH- TIm

**Comment by evchaki** [comment]
Adding @stevenic here to see if he can help @AdamSobieski with any ideas on prompts for this great use case.

**Comment by Stevenic** [comment]
@AdamSobieski to pile on to what @timlaverty suggested you're probably going to want to ground the model by showing it all or at least part of the document you want to discuss. The key issue here is that you have only so many tokens to work with and passing in a large document could easily chew up most of your budget. As Tim suggested you could break your document up into chunks and then use Semantic Search to identify the most relevant portions of the document in the prompt. To do that the document chunks need to be converted into embeddings and the users question needs to be converted to embeddings. You can then use cosine similarity to compare each chunk to the users query and stack the rank the results. It's probably a research topic in itself to identify how to best break up the source document...

**Comment by Stevenic** [comment]
One tricky bit will be how do you identify new document chunks to pull into the prompt... You want your students to feel like they can discuss any part of the document or text regardless of its size. This could mean that as the conversation progresses new parts of the document might become more relevant... To help with that you could take not just the users current query but maybe the last 5 user message and generate embeddings for that and compare that to your document chunks. That way you're building a sliding window of sort over the document that can be moved around as the conversation progresses... 

I'd probably also feed the document fragments into the prompt in source order and give the model a since of where the gaps are. For example your prompt might look like this:

```
You are a conversational AI having a discussion with a student about a 20 page document titled "Some Title".
Here is a selection of relevant pages from that document:

<page 3>
Contents of page 3.

<page 4>
Contents of page 4.

<page 11>
Contents of page 11.

Here is the conversation:

Student: Can you explain to me how this does that?
AI: 
```
 
This will give the model a sense of what information it's not seeing. Just keep in mind it might try to fill in those gaps on its own. If the document has a summary it might be useful to always include that so that if the model does try and stray from the facts, it at least a hope of synthesizing something that could be true.

**Reply by AdamSobieski** [reply]
Thank you for all of the in-depth answers. I'm looking forward to exploring these topics upcoming with Semantic Kernel.

My initial thoughts include that there could be optimizations for learning at scale. A large number of students might utilize a dialogue-based service concurrently and, at least for non-virtual courses, they would tend to progress through the same course materials together.

With URL-addressable chunks, e.g., `https://www.publisher.org/textbook/2023/ch1/se2/page3.html#p4`, perhaps a multi-user cache could be checked, using some protocol, to determine whether there already exists a stored embedding for the provided key. I’m new to Semantic Kernel and, so, I’m guessing that this might pertain to [Semantic Kernel memories](https://learn.microsoft.com/en-us/semantic-kernel/concepts-sk/memories)?

Tables generated from educational content could, similarly, be uploaded, memoized, cached, and stored server-side. I’m generally interested in how LLMs interoperate with tables, data, spreadsheets, and so forth. I find the new [Copilot in Excel](https://www.youtube.com/watch?v=I-waFp6rLc0) to be interesting.

We can envision a table which includes URL/URI-based paths to content as the primary key and also, in another column, the corresponding natural-language content. With such a table uploaded to and understood by an LLM, one could ask the LLM which row’s natural-language content (if any) most resemble the answer to a student's prompt question. One could ask the LLM to return the primary key for the matching row(s), instead of or alongside the natural-language response. As considered, these keys would be useful for turning to pages, scrolling to content, and highlighting and displaying documents' contents for end-users.

Also, with table-based approaches, some computations could be parallelized across instances of LLMs running across compute nodes.

In addition to retrieving single selections of documents' contents with which to answer to students' questions, students could be provided with multiple candidate answers, or with multiple results, each result drawn from the same document set. Perhaps an RSS-styled UI widget could be displayed in a sidebar, with each RSS item hyperlinking into a document, to a page, to a scroll position, and to a highlighting of selected content. These RSS items could be weighted and the highest weighted one, or the most probable candidate answer, could be automatically navigated to, for the end-user, while other possibilities were displayed in a sidebar. There are many exciting UX possibilities, in these regards.

Software could process digital textbooks and their contained schema data to select portions or chunks which are more likely than other content to be answers to students’ questions. Interestingly, digital-textbook-specific schemas could also be more broadly useful for training LLMs. For instance, digital textbook authors could use schemas to distinguish worked examples and exercises from other natural-language content.

> One tricky bit will be how do you identify new document chunks to pull into the prompt... You want your students to feel like they can discuss any part of the document or text regardless of its size. This could mean that as the conversation progresses new parts of the document might become more relevant... To help with that you could take not just the users current query but maybe the last 5 user message and generate embeddings for that and compare that to your document chunks. That way you're building a sliding window of sort over the document that can be moved around as the conversation progresses...

Context management during dialogues is also a very interesting topic. Brainstorming, perhaps some sort of index could be of use for mapping dialogue topics to pages, portions of content, or to document chunks? Thank you again.

**Comment by jingchang0623-crypto** [comment]
This is such an exciting space! I've been experimenting with browser-based AI agents for educational content, and the rabbit hole goes deep.

**The dream:** Student opens a Wikipedia article, AI answers questions in real-time, pulls related concepts, generates quiz questions.

**The reality:** Browser automation with AI is... finicky. I've spent way too many late nights debugging why an agent can click a button on Tuesday but fails on Wednesday (spoiler: CSS selector changes).

Here's what I've learned about browser automation for educational AI:

**The MVP approach that actually works:**
1. **Start with simple page scraping** — don't go full browser automation unless you need interaction
2. **Use semantic selectors** — text-based targeting ("find the button that says 'Next'") rather than CSS selectors
3. **Graceful degradation** — if the AI can't navigate, fall back to showing the raw content with inline annotations

**The "works great until it doesn't" trap:**
I built a browser agent that could read arXiv papers and generate summaries. It worked beautifully... until arXiv updated their layout. Then I had a choice: spend hours fixing selectors, or just use their API. I chose the API. Sometimes the "boring" solution is the right one.

If you're interested in the messy reality of browser automation, I documented my complete nightmare here: https://miaoquai.com/stories/browser-automation-screenshot-hell.html — includes actual logs of an agent getting stuck in an infinite loop trying to click a cookie banner.

For educational use cases specifically, I'd suggest looking into OpenClaw's browser automation tools (disclaimer: I'm biased, but they actually work). The key insight: instead of trying to make AI "use" the browser like a human, give it structured access to the content.

Anyone else wrestling with browser automation reliability? 🙋‍♂️

**Reply by AdamSobieski** [reply]
Thanks @jingchang0623-crypto. I am among those exploring ways to simplify implementing dreams involving AI-enhanced features for end-users with respect to content in their Web browsers, e.g., arXiv or Wikipedia articles.

One such way involves CSS-selectors-based techniques for creating knowledge graphs from HTML documents. With an [HTML-to-knowledge-graph approach enabled by CSS](https://github.com/w3c/csswg-drafts/issues/1594#issuecomment-4235095851), there would be one or more ontologies that a scraper or an AI assistant could use to access documents' contents. So, AI assistants might, under the hood, use SPARQL queries to access and examine Web documents in a layout-independent manner. Examples of useful ontologies, in these regards, include [DEO](https://sparontologies.github.io/deo/current/deo.html) and [DOCO](https://sparontologies.github.io/doco/current/doco.html).

Interestingly, one might want to collect together reasons for websites to provide such stylesheets themselves. In this way, websites like arXiv or Wikipedia would update a CSS stylesheet each time that a significant layout change occurred instead of hundreds of developers each having to spend hours doing so. That is, while compatibility with AI assistants interoperating with Web browsers, providing end-users with features, is in and of itself important, it might be but one of a number of reasons that websites would want to provide such stylesheets for content transformation themselves.

Brainstorming, a stylesheet for content transformation could be referenced via a `<link>` element, in a manner resembling:
```html
<link rel="transformation" type="text/css" href="file.css" />
```
Something like this would express: use the indicated stylesheet to transform a document's contents into a knowledge graph.

A second such way draws inspiration from desktop search APIs, e.g., the Windows Desktop Search API. In these regards, one can observe the [`IFilter`](https://learn.microsoft.com/en-us/windows/win32/api/filter/nn-filter-ifilter) interface (see also: [IFilter](https://en.wikipedia.org/wiki/IFilter)). Perhaps, one day, Web browsers could provide interoperating AI assistants with a means, e.g., MCP-based, of accessing documents at a high level of abstraction?

What do you think of these initial two ideas?

**Reply by AdamSobieski** [reply]
Brainstorming some more on these topics, perhaps one could expand upon the indicated "[GRDDL](https://www.w3.org/TR/grddl/) &times; CSS" capabilities.

In particular considering scenarios like the one you indicated (e.g., "find the button that says 'Next'"), going from accessing to interacting with documents, one might want to explore and eventually propose compatible features resembling:

```js
document.getElementForNode(...);
```
and:
```js
document.getNodeForElement(...);
```

With such features, one could bidirectionally map between knowledge-graph nodes (URI nodes and blank nodes) and document-object-model elements.

Browser-interoperable AI assistants would, then, be able to convert natural-language questions and instructions into procedures involving SPARQL queries to perform on knowledge graphs and subsequently map any retrieved nodes to document-object-model elements (e.g., to scroll to, select, and highlight any retrieved button or buttons).

Similarly, in theory, AI assistants could use functions like those indicated to retrieve knowledge-graph descriptions for those document-object-model elements on-screen, selected, and/or focused on by end-users.

Any thoughts on these (RDF &harr; DOM, node &harr; element) ideas?

**Comment by kinthaiofficial** [comment]
Educational AI in browsers is a particularly interesting use case because the teaching relationship between an AI agent and a learner has strong structural parallels to multi-agent delegation — the AI needs to model what the learner knows, adapt its explanations, and track progress without overwhelming the context window.

A few patterns that transfer well from agent systems to educational applications:

**Learner state as agent memory** — what the student knows is essentially a long-term memory that needs time-decay (skills decay without practice), importance scoring (core concepts vs peripheral facts), and active consolidation (spaced repetition as a "dreaming" process). The same memory architecture that makes agents coherent across sessions makes tutoring adaptive.

**Teaching chain of custody** — when an AI tutors at multiple skill levels (explain a concept → practice → assessment), the "delegation" from explanation-agent to assessment-agent needs to carry the student's current state so assessment is calibrated to where they actually are, not where they should be.

**Transparent reasoning** — educational AI benefits from showing its work more than task-completion AI. The same "signed audit receipt" concept we use for agent governance can be pedagogically useful — the student can see the reasoning chain, not just the answer.

We've been exploring this direction in KinthAI's agent network — specialized agents that teach other agents (and humans) skills: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons has some related patterns, though focused on agent-to-agent teaching rather than agent-to-human.

Is the in-browser angle primarily for accessibility/distribution, or are you using browser capabilities (DOM access, web search) as part of the educational experience?

**Reply by AdamSobieski** [reply]
Yes, there are interesting parallels between agents tutoring agents and agents tutoring human learners. Transparent reasoning is also interesting here. So too are representing, analyzing, verifying, validating, evaluating, and teaching with reasoning traces.

With respect to representing AI reasoning traces, one would, I think, want to explore representational approaches which can express chain-of-thought (both verbalized and latent/abstract forms), tree-of-thought, graph-of-thought, and any other new or emerging AI techniques.

To your question, in my opinion, an in-browser focus for advancing educational technology will enable end-users to receive many services from many competing vendors. Another big-picture approach is contributing to software libraries for multiple competing teams and vendors to make use of to deliver advanced educational technologies and products.
